File: C#/0147. Insertion Sort List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode InsertionSortList(ListNode head) {
        ListNode node=new ListNode(0);
        node.next=head;
        ListNode p=node;
        while(p.next!=null && p.next.next!=null){
            if(p.next.val<=p.next.next.val){
                p=p.next;
            }
            else{
                ListNode cur=p.next.next;
                p.next.next=cur.next;
                
                ListNode tmp=node;
                while(tmp.next.val<cur.val){
                    tmp=tmp.next;
                }
                cur.next=tmp.next;
                tmp.next=cur;
            }
        }
        return node.next;
    }
}

File: archives/20190519python/0147. Insertion Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        node=ListNode(0)
        node.next=head
        p=node
        while p.next and p.next.next:
            if p.next.val<=p.next.next.val:
                p=p.next
            else:
                cur=p.next.next
                p.next.next=cur.next
                #p=p.next
                tmp=node
                while tmp.next.val<cur.val:
                    tmp=tmp.next
                cur.next=tmp.next
                tmp.next=cur
        return node.next
